### DISCUSSION TITLE
How to pass c# action to python script and call it from there

### DISCUSSION THREAD

**Discussion by NataliVol4ipsa** [pythonnet/pythonnet#1962] [Q&A] [answered]
Hi! Im using latest version of pythonnet and .net6.
I have a UI c# app and want to be able to execute long-running (~30 minutes) python script from it. I need to show progress bar and handle some notifications sent by python script. Of course, the script has all the info about its progress, it just needs to call some function, so that c# app is also aware that something happens. How can this be implemented?

For example, in IronPython I would create
`dynamic proxy = new ExpandoObject();`
then add my handling delegate
`proxy.ShowMessage = new Action<string>(ShowMessage);`
where ShowMessage is some method that interacts with UI. And eventually I would add proxy to the scope
`scope.SetVariable("proxy", CreateProxy());`

then python script code would look like this
`proxy.ShowMessage("Hello from script")` and will work great.

however, if I use this approach with pythonnet, like this
`scope.Set("proxy", CreateProxy().ToPython());`
`scope.Exec("proxy.ShowMessage('Test')");`
then I get an exception
Python.Runtime.PythonException: ''ExpandoObject' object has no attribute 'ShowMessage''
Is it me doing this wrong? Or should I use different approach? Which one?

I have checked issues, discussions, and mailing list, but didnt find the solution. I saw some crazy guy trying to override c# properties functionality so that they raise events, and that feels like an overkill for such an issue. So I hope that there is a better solution. 

Thanks in advance!

**Comment by lostmsu** [comment, marked as answer]
1. Python.NET does not support `ExpandoObject`. You need to define a C# class with method or property named `ShowMessage`.
2. Alternatively, you can replace `ExpandoObject` with `PyObject`.
3. You actually don't need an entire class or event an instance of object with `ShowMessage` attribute. You can directly do `scope.SetVariable("show_message", new Action<string>(ShowMessage))` and then in Python `show_message('42')`

**Reply by NataliVol4ipsa** [reply]
Works like a charm! I have been struggling with this one for so long, and solution ended up being so simple! Thank you very much, you are my savior :)

**Reply by GlenHerman** [reply]
I need help with doing the same thing.  In the response, scope.SetVariable is used, but in pythonnet, I don't find setVariable, just Set.
